### PR TITLE
feat(scene composer): icon picker rule changes

### PR DESCRIPTION
**Pull request by divya-sea**
## Overview
This contains following changes,
1. Added icon rule changes.
2. Moved customColors under properties.
3. Added unit tests.

## Verifying Changes

https://github.com/awslabs/iot-app-kit/assets/54058998/c4344955-ef79-4a9b-a733-2c6eb0ff4e1e

### Scene Composer
For `scene-composer` package changes specifically, you can preview the component in the published storybook artifact. To do this, wait for the `Publish Storybook` action to complete below.

- Click on the workflow details
- Select the Summary item on the left
- Download the zip file

To run the storybook build locally, you need a local static web server:

```
npm install -g httpserver
cd <Extracted Zip Directory>
httpserver
```

Then open the website http://localhost:8080 to run the doc site.

## Legal
This project is available under the [Apache 2.0 License](http://www.apache.org/licenses/LICENSE-2.0.html).

### CODE REVIEW

**Comment by sheilaXu** [comment]
from the screen recording I saw a couple of issues:
1. after you set the data binding and rule id for the tag component, it disappears for a while (~0.07 minute) until the rule has the icon target set. this is not a good experience and the tag component should just ignore the rule when it's not for icon instead of disappearing
2. the first rule expression in the rules panel displays an `info` icon, but when the rule is applied, the tag still displays the `3` icon from the default style. I know you display the `info` icon as default value for icon picker, but it appears to me like the icon from the rule is not applied to the tag. I suggest to not display default icon in the picker to avoid confusion

File: packages/scene-composer/src/utils/dataBindingUtils.ts [hunk 1]
```diff
@@ -160,25 +160,22 @@ export const ruleEvaluator = (
...
+        return { target: statement.target, targetMetadata: statement.targetMetadata };
       }
     } catch (error) {
       // It might be a user mistake for one rule, but the next rule may work, we should continue
       LOG.error('Failed to evaluate rule', error as Error);
     }
   }
-
-  return defaultState;
+  return { target: defaultState, targetMetadata: undefined };
```
Comment (sheilaXu) [inline packages/scene-composer/src/utils/dataBindingUtils.ts:180]: can you add a return type to this function so the consumers know clearly the type of the value returned? right now I can see 2 returns `{ target: xxx, targetMetadata: yyy}` but the rest return different types. it will be hard to handle the results in this way.
also instead of `{ target: xxx, targetMetadata: yyy}`, I suggest `{ result: xxx, targetMetadata: yyy}` so it makes more sense for cases returning defaultState or number.

File: packages/scene-composer/src/interfaces/dataTypes.ts [hunk 2]
```diff
@@ -34,6 +34,11 @@ export interface ITransformConstraint {
 }
 
 export type { IValueDataBinding, ITwinMakerEntityDataBindingContext } from '@iot-app-kit/source-iottwinmaker';
+export type TargetMetadata<T> = Record<string, T> & {
+  color: string;
+  iconPrefix: string;
+  iconName: string;
```
Comment (sheilaXu) [inline packages/scene-composer/src/interfaces/dataTypes.ts:40]: I guess they should all be optional

File: packages/scene-composer/src/interfaces/dataTypes.ts [hunk 3]
```diff
@@ -34,6 +34,11 @@ export interface ITransformConstraint {
 }
 
 export type { IValueDataBinding, ITwinMakerEntityDataBindingContext } from '@iot-app-kit/source-iottwinmaker';
+export type TargetMetadata<T> = Record<string, T> & {
```
Comment (sheilaXu) [inline packages/scene-composer/src/interfaces/dataTypes.ts:37]: maybe remove `Record<string, T> &`  part and make this type more specific, so we have better typing support to avoid typos. 
I've seen problem with several other fields typed as `Record<string, xxx>` that we can hardly know the values being stored. so I think we should start to type with the clear keys and extend in the future when we need more keys, instead of starting with a generic map type

File: packages/scene-composer/src/components/panels/scene-rule-components/SceneRuleTargetEditor.tsx [hunk 4]
```diff
@@ -92,23 +102,47 @@ export const SceneRuleTargetEditor: React.FC<ISceneRuleTargetEditorProps> = ({
...
+                }}
+                selectedIcon={icon as IconLookup}
+                iconPickerLabel={formatMessage({ defaultMessage: 'Icon', description: 'Icon' })}
+                iconFilterText={formatMessage({ defaultMessage: 'Find icons', description: 'Find icons' })}
+                iconFilterTextAriaLabel={formatMessage({
+                  defaultMessage: 'Filter icons',
+                  description: 'Filter icons',
+                })}
+              />
+            </>
```
Comment (sheilaXu) [inline packages/scene-composer/src/components/panels/scene-rule-components/SceneRuleTargetEditor.tsx:145]: all of those are specific to icon target, so they should be in `SceneRuleTargetIconEditor` component instead of here. and all states related to chosenColor and customIcon should all be move into `SceneRuleTargetIconEditor`

File: packages/scene-composer/src/components/panels/scene-components/tag-style/IconPicker/IconPicker.tsx [hunk 5]
```diff
@@ -108,14 +107,14 @@ export const IconPicker = ({
                 />
                 <div style={IconGridLayout}>
                   {filteredIcons.map((icon, index) => {
-                    return <FontAwesomeIcon key={index} icon={icon} onClick={() => handleIconClick(icon)} />;
+                    return <FontAwesomeIcon key={index + 1} icon={icon} onClick={() => handleIconClick(icon)} />;
```
Comment (sheilaXu) [inline packages/scene-composer/src/components/panels/scene-components/tag-style/IconPicker/IconPicker.tsx:110]: why is `+ 1`?

File: packages/scene-composer/src/components/panels/SceneRulesPanel.spec.tsx [hunk 6]
```diff
@@ -1,5 +1,6 @@
 import React from 'react';
-import { render } from '@testing-library/react';
+import { fireEvent, render } from '@testing-library/react';
+import wrapper from '@awsui/components-react/test-utils/dom';
```
Comment (sheilaXu) [inline packages/scene-composer/src/components/panels/SceneRulesPanel.spec.tsx:3]: are there unused imports?

File: packages/scene-composer/src/components/panels/scene-components/AnchorComponentEditor.tsx [hunk 7]
```diff
@@ -216,8 +219,11 @@ export const AnchorComponentEditor: React.FC<IAnchorComponentEditorProps> = ({
                   chosenColor: pickedColor,
                 });
               }}
-              onUpdateCustomColors={(chosenCustomColors) => onUpdateCallback({ customColors: chosenCustomColors })}
-              customColors={anchorComponent.customColors}
+              onUpdateCustomColors={(chosenCustomColors) =>
+                setSceneProperty(KnownSceneProperty.TagCustomColors, chosenCustomColors)
+              }
+              // onUpdateCallback({ customColors: chosenCustomColors })}
```
Comment (sheilaXu) [inline packages/scene-composer/src/components/panels/scene-components/AnchorComponentEditor.tsx:225]: clean up?

File: packages/scene-composer/src/augmentations/components/three-fiber/anchor/AnchorWidget.tsx [hunk 8]
```diff
@@ -113,16 +106,57 @@ export function AsyncLoadedAnchorWidget({
...
+      // check overrideCustomColor is undefined but targetMetadata color is defined
+      if (!overrideCustomColor && ruleResult.targetMetadata?.color) {
+        setOverrideCustomColor(ruleResult.targetMetadata?.color);
+      }
+
+      // check overrideCustomIcon is defined but not same as rule targetMetadata icon
+      if (
+        overrideCustomIcon &&
+        overrideCustomIcon.iconName !== ruleResult.targetMetadata?.iconName &&
+        overrideCustomIcon.prefix !== ruleResult.targetMetadata?.iconPrefix
```
Comment (sheilaXu) [inline packages/scene-composer/src/augmentations/components/three-fiber/anchor/AnchorWidget.tsx:127]: this should be `||` because icon prefix can be unchanged, and only icon name is changed. we should also update in this case

File: packages/scene-composer/src/augmentations/components/three-fiber/anchor/AnchorWidget.tsx [hunk 9]
```diff
@@ -113,16 +106,57 @@ export function AsyncLoadedAnchorWidget({
   useEffect(() => {
     setParent(node.parentRef ? getObject3DFromSceneNodeRef(node.parentRef) : undefined);
   }, [node.parentRef]);
-
   // Evaluate visual state based on data binding
   const visualState = useMemo(() => {
-    getCustomIconColor(ruleResult, setOverrideCustomColor);
-    const ruleTargetInfo = getSceneResourceInfo(ruleResult as string);
+    const checkTargetExists = typeof ruleResult === 'object' && 'target' in ruleResult;
+    const checkTargetMetadataExists = typeof ruleResult === 'object' && 'targetMetadata' in ruleResult;
```
Comment (sheilaXu) [inline packages/scene-composer/src/augmentations/components/three-fiber/anchor/AnchorWidget.tsx:112]: with the property return type defined for evaluateRule function, we don't need those check

File: packages/scene-composer/src/augmentations/components/three-fiber/anchor/AnchorWidget.tsx [hunk 10]
```diff
@@ -113,16 +106,57 @@ export function AsyncLoadedAnchorWidget({
...
+        overrideCustomIcon.prefix !== ruleResult.targetMetadata?.iconPrefix
+      ) {
+        setOverrideCustomIcon({
+          prefix: ruleResult.targetMetadata?.iconPrefix,
+          iconName: ruleResult.targetMetadata?.iconName,
+        } as IconLookup);
+      }
+
+      // check overrideCustomIcon is undefined but targetMetadata icon is defined
+      if (ruleResult.targetMetadata && ruleResult.targetMetadata.iconName && ruleResult.targetMetadata.iconPrefix) {
```
Comment (sheilaXu) [inline packages/scene-composer/src/augmentations/components/three-fiber/anchor/AnchorWidget.tsx:136]: from the comment, you also want condition `!overrideCustomIcon`?

File: packages/scene-composer/src/augmentations/components/three-fiber/anchor/AnchorWidget.tsx [hunk 11]
```diff
@@ -113,16 +106,57 @@ export function AsyncLoadedAnchorWidget({
...
+          prefix: ruleResult.targetMetadata?.iconPrefix,
+          iconName: ruleResult.targetMetadata?.iconName,
+        } as IconLookup);
+      }
+    }
+    const ruleTargetInfo = getSceneResourceInfo(
+      checkTargetExists ? (ruleResult.target as string) : (ruleResult as string),
+    );
     // Anchor widget only accepts icon, otherwise, default to Info icon
     return ruleTargetInfo.type === SceneResourceType.Icon ? ruleTargetInfo.value : defaultIcon;
```
Comment (sheilaXu) [inline packages/scene-composer/src/augmentations/components/three-fiber/anchor/AnchorWidget.tsx:147]: all of above logic should run only when `ruleTargetInfo.type === SceneResourceType.Icon` is true. otherwise return default icon

File: packages/scene-composer/src/interfaces/dataTypes.ts (hunk 3 above)
Comment (divya-sea) [reply packages/scene-composer/src/interfaces/dataTypes.ts:37]: In our case T is color,icon and iconName and in the future if you all together want to create TargetMetadata of different type then this record still holds good. I guess we can keep as it is but internally color, icon and iconName we can specify a type. I would like to keep this generic type as it is but T type can be specified internally.

File: packages/scene-composer/src/components/panels/scene-rule-components/SceneRuleTargetEditor.tsx (hunk 4 above)
Comment (divya-sea) [reply packages/scene-composer/src/components/panels/scene-rule-components/SceneRuleTargetEditor.tsx:145]: They are specific to Icon and logic is within it. There are two reasons why we need to add them here. 
1. Because of the current grid pattern. There is an outer most grid for targets.
2. There is grid within target components itself
If we have move them the inside icon component then we will have change the UX design as there is a outer grid structure which set the icon component to render from colSpan 4 If you add them part of the icon component. 
I would like to keep as it is for this release unless there is an urgent change in UI requirement.

File: packages/scene-composer/src/utils/dataBindingUtils.ts (hunk 1 above)
Comment (divya-sea) [reply packages/scene-composer/src/utils/dataBindingUtils.ts:180]: There are total 5 return results here. We can return targetMetadata if the type of target is icon. But outermost return will return the default target state if nothing is matched but I just added undefined because we need this to evaluate visual state.

File: packages/scene-composer/src/augmentations/components/three-fiber/anchor/AnchorWidget.tsx (hunk 11 above)
Comment (divya-sea) [reply packages/scene-composer/src/augmentations/components/three-fiber/anchor/AnchorWidget.tsx]: yes. ad also filtered the icons in anchor component as well since we have support only for icon.

File: packages/scene-composer/src/components/panels/scene-components/AnchorComponentEditor.tsx (hunk 7 above)
Comment (divya-sea) [reply packages/scene-composer/src/components/panels/scene-components/AnchorComponentEditor.tsx:244]: done

File: packages/scene-composer/src/components/panels/SceneRulesPanel.spec.tsx (hunk 6 above)
Comment (divya-sea) [reply packages/scene-composer/src/components/panels/SceneRulesPanel.spec.tsx:3]: done

File: packages/scene-composer/src/utils/dataBindingUtils.ts (hunk 1 above)
Comment (divya-sea) [reply packages/scene-composer/src/utils/dataBindingUtils.ts:190]: Returning a RuleReturnResult type.

File: packages/scene-composer/src/components/panels/SceneRulesPanel.spec.tsx (hunk 6 above)
Comment (sheilaXu) [reply packages/scene-composer/src/components/panels/SceneRulesPanel.spec.tsx:3]: this doesn't seem to be addressed

File: packages/scene-composer/package.json [hunk 12]
```diff
@@ -150,6 +150,7 @@
     "@react-three/postprocessing": "2.6.2",
     "@react-three/test-renderer": "^8.1.3",
     "@tanstack/react-query": "^4.29.15",
+    "@testing-library/user-event": "^14.4.3",
```
Comment (sheilaXu) [inline packages/scene-composer/package.json:153]: this new dependency doesn't seem to be used anywhere. please remove it if so

File: packages/scene-composer/src/components/panels/scene-components/AnchorComponentEditor.tsx (hunk 7 above)
Comment (sheilaXu) [reply packages/scene-composer/src/components/panels/scene-components/AnchorComponentEditor.tsx:244]: change not committed here as well

File: packages/scene-composer/src/components/panels/scene-components/tag-style/ColorPicker/ColorPicker.tsx [hunk 13]
```diff
@@ -70,12 +72,15 @@ export const ColorPicker = ({
...
-    if (!Object.values(palleteColors).includes(color)) {
-      setCustomInternalColors(customInternalColors.concat(color));
-      onUpdateCustomColors?.([...new Set(customInternalColors)]);
-    }
-  };
+  const checkIfCustomColor = useCallback(
+    (color): void => {
+      if (!Object.values(palleteColors).includes(color) && color !== colors.infoBlue) {
+        setCustomInternalColors(customInternalColors.concat(color));
+        onUpdateCustomColors?.([...new Set(customInternalColors)]);
```
Comment (sheilaXu) [inline packages/scene-composer/src/components/panels/scene-components/tag-style/ColorPicker/ColorPicker.tsx:79]: `customInternalColors` here is not updated to have the new color yet since setState is async. you should have a local array with the new value

File: packages/scene-composer/src/hooks/useRuleResult.ts [hunk 14]
```diff
@@ -1,13 +1,18 @@
...
 import { useSceneDocument, useDataStore } from '../store';
 import { useSceneComposerId } from '../common/sceneComposerIdContext';
 import { dataBindingValuesProvider, ruleEvaluator } from '../utils/dataBindingUtils';
 
 import useBindingData from './useBindingData';
 
+export interface RuleReturnResult {
+  target: string | number;
+  targetMetadata?: TargetMetadata<string>;
+}
```
Comment (sheilaXu) [inline packages/scene-composer/src/hooks/useRuleResult.ts:14]: this type should be defined in dataBindingUtils file because `useRuleResult` imports from `dataBindingUtils`. if you add it here, `dataBindingUtils` imports from `useRuleResult` and creates the circular dependency

File: packages/scene-composer/src/components/three-fiber/MotionIndicatorComponent/__tests__/__snapshots__/MotionIndicatorComponentSnap.spec.tsx.snap [hunk 15]
```diff
@@ -89,7 +89,7 @@ exports[`MotionIndicatorComponent should render with correct speed and values fr
     <div
       data-testid="linear-plane"
     >
-      [{"speed":3,"foregroundColor":0,"backgroundColor":0,"config":{"numOfRepeatInY":2,"defaultForegroundColor":"red","defaultBackgroundColor":"blue","defaultSpeed":6,"backgroundColorOpacity":1},"scale":[1,2,3]},{}]
+      [{"speed":0,"foregroundColor":0,"backgroundColor":0,"config":{"numOfRepeatInY":2,"defaultForegroundColor":"red","defaultBackgroundColor":"blue","defaultSpeed":6,"backgroundColorOpacity":1},"scale":[1,2,3]},{}]
```
Comment (sheilaXu) [inline packages/scene-composer/src/components/three-fiber/MotionIndicatorComponent/__tests__/__snapshots__/MotionIndicatorComponentSnap.spec.tsx.snap:92]: why is this changed? something seems to be wrong

File: packages/scene-composer/src/interfaces/dataTypes.ts (hunk 3 above)
Comment (sheilaXu) [reply packages/scene-composer/src/interfaces/dataTypes.ts:37]: I don't think a set of metadata will always have the same type, so the generic type here will not be super useful here. and also if we need more kinds of metadata, we should defined in here to make it more clear. so the generic `Record<>` part will not be useful as well

File: packages/scene-composer/src/components/panels/scene-rule-components/SceneRuleTargetIconEditor.tsx [hunk 16]
```diff
@@ -38,38 +42,43 @@ export const SceneRuleTargetIconEditor: React.FC<ISceneRuleTargetIconEditorProps
...
-            'Specifies the localized string that describes an option as being selected. This is required to provide a good screen reader experience',
-        })}
-      />
-      {isCustomStyle ? (
-        <DecodeSvgString
-          selectedColor={chosenColor ?? colors.customBlue}
-          iconString={iconString}
-          width='32px'
-          height='32px'
+    <>
```
Comment (sheilaXu) [inline packages/scene-composer/src/components/panels/scene-rule-components/SceneRuleTargetIconEditor.tsx:48]: this layer doesn't seem to be needed

File: packages/scene-composer/src/components/panels/scene-rule-components/SceneRuleTargetEditor.tsx [hunk 17]
```diff
@@ -92,23 +102,47 @@ export const SceneRuleTargetEditor: React.FC<ISceneRuleTargetEditorProps> = ({
...
+              <IconPicker
+                onSelectIconChange={(pickedIcon) => {
+                  onChange(convertToIotTwinMakerNamespace(targetInfo.type, targetInfo.value), {
+                    ...targetMetadata,
+                    iconPrefix: pickedIcon.prefix,
+                    iconName: pickedIcon.iconName,
+                  } as TargetMetadata<string>);
+                  setIcon(pickedIcon as IconLookup);
+                }}
+                selectedIcon={icon as IconLookup}
```
Comment (sheilaXu) [inline packages/scene-composer/src/components/panels/scene-rule-components/SceneRuleTargetEditor.tsx:137]: do you need to cast to `IconLookup` here? can we use `IIconLookup` for any internal component, then only cast to `IconLookup` when passing it to fontawsome?

File: packages/scene-composer/package.json (hunk 12 above)
Comment (divya-sea) [reply packages/scene-composer/package.json:153]: We can use it in the future instead of fireEvent. This is better version of fireEvent.

File: packages/scene-composer/src/interfaces/dataTypes.ts (hunk 3 above)
Comment (divya-sea) [reply packages/scene-composer/src/interfaces/dataTypes.ts:37]: You meant to they will have the same type? If it is not of same then, T holds good. All you have to do is 
```
color?: string;
 iconPrefix?: string;
 iconName?: string;
opacity?: Object;
meta?: string[];
```
If we string just to Record<string, string>, the above case does not work. That time again we will have to create a new type.

File: packages/scene-composer/src/hooks/useRuleResult.ts (hunk 14 above)
Comment (divya-sea) [reply packages/scene-composer/src/hooks/useRuleResult.ts:14]: Yeah I wanted to know what is the best place to declare this type.

File: packages/scene-composer/src/components/three-fiber/MotionIndicatorComponent/__tests__/__snapshots__/MotionIndicatorComponentSnap.spec.tsx.snap (hunk 15 above)
Comment (divya-sea) [reply packages/scene-composer/src/components/three-fiber/MotionIndicatorComponent/__tests__/__snapshots__/MotionIndicatorComponentSnap.spec.tsx.snap:92]: I can run them again.check what changed this.

File: packages/scene-composer/src/components/panels/scene-rule-components/SceneRuleTargetIconEditor.tsx (hunk 16 above)
Comment (divya-sea) [reply packages/scene-composer/src/components/panels/scene-rule-components/SceneRuleTargetIconEditor.tsx:48]: yes We don't need this.

File: packages/scene-composer/src/components/panels/scene-rule-components/SceneRuleTargetEditor.tsx (hunk 17 above)
Comment (divya-sea) [reply packages/scene-composer/src/components/panels/scene-rule-components/SceneRuleTargetEditor.tsx:137]: No we don't

File: packages/scene-composer/package.json (hunk 12 above)
Comment (sheilaXu) [reply packages/scene-composer/package.json:153]: we can add it in the future when it's actually in use. I don't see a reason to add a library in advance unless you know you will use it in next change

Comment (divya-sea) [reply packages/scene-composer/package.json:153]: we should not be using firevent actually. It is out dated. We should be using this and replacing fireevent.

File: packages/scene-composer/src/components/three-fiber/MotionIndicatorComponent/MotionIndicatorComponent.tsx [hunk 18]
```diff
@@ -72,8 +72,10 @@ const MotionIndicatorComponent: React.FC<IMotionIndicatorComponentProps> = ({
         dataBindingTemplate,
       );
 
-    const result = ruleEvaluator(0, values, speedRule) as number;
-    return result >= 0 ? result : 0;
+    const result = ruleEvaluator(0, values, speedRule);
```
Comment (sheilaXu) [inline packages/scene-composer/src/components/three-fiber/MotionIndicatorComponent/MotionIndicatorComponent.tsx:75]: shouldn't it be `ruleEvaluator(0, values, speedRule).target`?

File: packages/scene-composer/src/interfaces/dataTypes.ts (hunk 3 above)
Comment (sheilaXu) [reply packages/scene-composer/src/interfaces/dataTypes.ts:37]: so whenever you add new metadata, you will add the field explicitly here. there isn't a case we will use `Record<string, T>`. it just make us unsure about the exact metadata can be in the metadata

File: packages/scene-composer/src/components/three-fiber/MotionIndicatorComponent/MotionIndicatorComponent.tsx (hunk 18 above)
Comment (divya-sea) [reply packages/scene-composer/src/components/three-fiber/MotionIndicatorComponent/MotionIndicatorComponent.tsx:75]: We are returning returning the RuleReturnResult and there is only single value in it. This is also one way of getting the returned result?

File: packages/scene-composer/src/interfaces/dataTypes.ts (hunk 3 above)
Comment (sheilaXu) [reply packages/scene-composer/src/interfaces/dataTypes.ts:37]: I've recently changed some types defined as `Record<xx, yy>` to be explicit interfaces, because I'm seeing problem for us to know the data stored without searching in the entire code base

File: packages/scene-composer/src/components/three-fiber/MotionIndicatorComponent/MotionIndicatorComponent.tsx (hunk 18 above)
Comment (sheilaXu) [reply packages/scene-composer/src/components/three-fiber/MotionIndicatorComponent/MotionIndicatorComponent.tsx:75]: but `ruleEvaluator` returns `{ target: xx, targetMetadata: yy}`, why would `typeof result === 'number'` be true?

Comment (divya-sea) [reply packages/scene-composer/src/components/three-fiber/MotionIndicatorComponent/MotionIndicatorComponent.tsx:75]: I removed targetMetadata from this. That is not required here. If (evalResult) type is number  we are returning Number(evalResult) so here in this rule evaluator speed is a number.

**Review by sheilaXu — APPROVED** [review]
